src/app/utils/immutable-array-fns.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * This utility file contains immutable array functions.
+ */
+
+/**
+ * Pure function to reorders given array from given `fromIndex` to `toIndex`.
+ * The new reordered array will be returned.
+ *
+ * Function retrieved from
+ * https://github.com/granteagon/move/blob/master/src/index.js and converted to
+ * TypeScript and renamed for clarity.
+ *
+ * @param array initial array to reorder
+ * @param fromIndex the current index of the element to move
+ * @param toIndex the new index to move the element to
+ * @returns reordered array
+ */
+export const reorder = <T>(
+  array: T[],
+  fromIndex: number,
+  toIndex: number,
+): T[] => {
+  const elementToMove = array[fromIndex]
+  const diff = fromIndex - toIndex
+
+  if (diff > 0) {
+    // Reorder to the left.
+    return [
+      ...array.slice(0, toIndex),
+      elementToMove,
+      ...array.slice(toIndex, fromIndex),
+      ...array.slice(fromIndex + 1, array.length),
+    ]
+  } else if (diff < 0) {
+    // Reorder to the right.
+    const targetIndex = toIndex + 1
+    return [
+      ...array.slice(0, fromIndex),
+      ...array.slice(fromIndex + 1, targetIndex),
+      elementToMove,
+      ...array.slice(targetIndex, array.length),
+    ]
+  }
+  return array
+}
+
+/**
+ * Pure function to replace element at given `index` with `newValue`.
+ *
+ * @param array initial array to replace element for
+ * @param index index to replace element at
+ * @param newValue the new value to replace with
+ *
+ * @return new array with replaced value
+ */
+export const replaceAt = <T>(array: T[], index: number, newValue: T): T[] => {
+  const ret = array.slice(0)
+  ret[index] = newValue
```

Comment:
again, this can result in `undefined`s in the middle of the array, but in this case I think this behaviour is expected

Reply:
alternatively, you might even consider making `findIndex` the responsibility of this function instead of the calling function

Reply:
Yea, probably not going to change this function as it does what it does. Probably the onus of the caller to check if the given index is valid or invalid.